plugin/trino-exchange/src/main/java/io/trino/plugin/exchange/FileSystemExchangeSink.java

```diff
@@ -0,0 +1,241 @@
+            }
+        }
+        else {
+            return NOT_BLOCKED;
+        }
+    }
+
+    @Override
+    public void add(int partitionId, Slice data)
+    {
+        throwIfFailed();
+
+        checkArgument(partitionId < outputPartitionCount, "partition id is expected to be less than %s: %s", outputPartitionCount, partitionId);
+        checkState(!finished, "already finished");
+        if (finished) {
+            return;
+        }
+
+        writers.computeIfAbsent(partitionId, this::createWriter);
+        synchronized (writers.get(partitionId)) {
+            writeToExchangeStorage(partitionId, Slices.wrappedIntArray(data.length()));
+            writeToExchangeStorage(partitionId, data);
+        }
+    }
+
+    private ExchangeStorageWriter createWriter(int partitionId)
+    {
+        URI outputPath = outputDirectory.resolve(partitionId + DATA_FILE_SUFFIX);
+        try {
+            return exchangeStorage.createExchangeStorageWriter(outputPath, secretKey);
+        }
+        catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    private void writeToExchangeStorage(int partitionId, Slice slice)
+    {
+        int position = 0;
+        while (position < slice.length()) {
+            SliceOutput pendingBuffer = pendingBuffers.computeIfAbsent(partitionId, ignored -> {
+                try {
+                    return freeBuffers.take();
+                }
+                catch (InterruptedException e) {
+                    Thread.currentThread().interrupt();
+                    throw new RuntimeException(e);
+                }
+            });
+            int writableBytes = min(pendingBuffer.writableBytes(), slice.length() - position);
+            pendingBuffer.writeBytes(slice.getBytes(position, writableBytes));
+            position += writableBytes;
+
+            flushIfNeeded(partitionId, false);
+        }
+    }
+
+    private void flushIfNeeded(int partitionId, boolean finished)
+    {
+        SliceOutput buffer = pendingBuffers.get(partitionId);
+        if (!buffer.isWritable() || finished) {
+            if (!buffer.isWritable()) {
+                pendingBuffers.remove(partitionId);
+            }
+            writers.get(partitionId).write(buffer.slice()).whenComplete((result, throwable) -> {
+                if (throwable != null) {
+                    failure.compareAndSet(null, throwable);
+                }
+
+                buffer.reset();
+                freeBuffers.add(buffer);
+
+                CompletableFuture<?> completableFuture;
+                synchronized (this) {
+                    completableFuture = blockedFuture;
+                }
+                if (!completableFuture.isDone()) {
+                    completableFuture.complete(null);
+                }
+            });
+        }
+    }
+
+    @Override
+    public long getMemoryUsage()
+    {
+        if (finished) {
+            return INSTANCE_SIZE;
```

Comment:
where are we dropping buffers so we can no longer account for them?

Reply:
Currently, when you call `finish`, we wait for all write futures to complete, or you call `abort`, we cancel all write futures. All the buffers will be added back to `freeBuffers`. Then, we call `clear` on `pendingBuffers`, `freeBuffers` and `writers`, and memory is reclaimed.

Reply:
For `finish`: all buffers are added back to `freeBuffers` queue, and later cleared
For `abort`: all in-progress writes are aborted, and we naturally lose reference of these buffers and they should be garbage collected.